### ISSUE TITLE
Latest Math-MPFR (version 4.31) fails one test on Strawberry Perl-5.40.0

### ISSUE THREAD

**Issue by sisyphus** [StrawberryPerl/Perl-Dist-Strawberry#226]
On SP-5.40.0:
```
> perl -MMath::MPFR=":mpfr" -le "Rmpfr_printf('%a', 3.0625);"
repl-vsnprintf.c:388: GNU MP assertion failed: len < total_width
```
The bug also occurs with Rmpfr_sprintf  .... and probably (untested) Rmpfr_fprintf and Rmpfr_snprintf.
This bug, where present, will not be found by the test suite of earlier versions of Math::MPFR for the simple reason that the test suite of those earlier version did not look for it.

The MPFR source's <code>INSTALL</code> file states:
```
In order to have the MPFR formatted output functions based on an
ISO-conforming printf(), you need to compile GMP (not MPFR) with
CC="gcc -D__USE_MINGW_ANSI_STDIO" (since the standard printf modifiers
%e, %Ld and %td are passed to GMP).
```
I don't think SP has ever passed that <code>CC="gcc -D__USE_MINGW_ANSI_STDIO"</code> argument to  <code>configure</code> when building the GMP library - but that hasn't always mattered.
For example, the bug is not present on SP-5.38.2, but only because an MSVCRT compiler was used.
Nor is it present on  SP-5.36.1, but it is present on 5.32.1 and earlier.

However, now that a UCRT compiler is being used, that INSTALL file directive needs to be followed.
(I, too, had been ignoring that directive on some of my builds, oblivious to this issue until a couple of days ago.)

With affected builds of Math::MPFR, the bug is fairly easy to work around, as the problem arises only with "%a" and "%A" formatting, and only when Rmpfr_printf() is attempting to format a plain perl scalar (IV or NV).
One can just do <code>printf('%a', 3.0625)</code> instead. (Why use Rmpfr_printf when printf() will do the job ?)

The main purpose of Rmpfr_printf is to format values held by Math::MPFR objects, and such formatting is not subject to this bug.
On SP-5.40.0:
```
perl -MMath::MPFR=":mpfr" -le "Rmpfr_printf('%Ra', Math::MPFR->new(3.0625));"
0x3.1p+0
```

I think that Math::GMP doesn't plug into the <code>gmp_printf()</code> functionality and is thus completely unaffected by this.

AFAICT gmp_printf() works reliably, irrespective of whether we build GMP with or without <code>CC="gcc -D__USE_MINGW_ANSI_STDIO"</code>.
If that argument is omitted, then configure might not realize that 'vsnprintf' is available - in which case it will use the code in <code>repl_vsnprintf.c</code>, but I think that only results in some loss of efficiency.
AFAICT, the repl_vsnprintf.c code works reliably for GMP, and it's only MPFR that can't handle it properly.

If the GMP library is built on windows with that recommended configure arg, then we are assured that the code in repl_vsnprintf.c</code> will not be called upon.

**Comment by shawnlaffan** [comment]
Yes, the build for GMP does not include that flag:

https://github.com/StrawberryPerl/build-extlibs/blob/d22291c1a0de960466d1983cb0af88bff8c974a7/build.sh#L611-L626

**Comment by sisyphus** [comment]
> One can just do printf('%a', 3.0625) instead. (Why use Rmpfr_printf when printf() will do the job ?)

Seems that "%a" formatting wasn't available to perl until version 5.22.
A more widely portable workaround would therefore be to instead call <code>Rmpfr_printf('%Ra', Math::MPFR->new(3.0625));</code>

I'll fix Math::MPFR to automatically (at module load time) engage that workaround if (and only if) such a buggy mpfr library is being used.

**Comment by sisyphus** [comment]
Fixed in Math-MPFR-4.32.
